docs/antrea-l7-network-policy.md

```diff
@@ -362,6 +364,30 @@ Deny ingress from client (10.10.1.5) to web (10.10.1.4/admin)
 }
 ```
 
+### L7 Visibility
+
+Layer 7 network flow visibility enables users to see the application layer flows in the network and
+visualizing them on Theia.
+
+#### Usage
+
+To achieve L7 (Layer 7) network flow visibility, feature gate, namely `L7Visibility` must be enabled.
+User could activate the L7 network visibility feature either by applying L7 NetworkPolicies as
```

Comment:
It's been a while since my last review, but I am not sure I understand why applying an L7 NetworkPolicy on a Pod automatically generates L7 data to be included in the flow data. This functionality seems to be driven by the implementation. From a user perspective, IMO it makes more sense to have a single mechanism for collecting the L7 data (annotation of Pod / Namespace). @tnqn what do you think?

Reply:
My point was that even if an L7NetworkPolicy is applied to a Pod and we get L7 visibility information "for free", we should not include it in the data sent to the FlowAggregator, unless L7 visibility is also enabled on the Pod.

Reply:
I see the point of using single mechanism, user would only be able to visualize the L7 flows by enabling `L7Visibility` featureGate regardless of `L7NetworkPolicy` enable state.

- Current Implementation: 
 User enables `L7NetworkPolicy` feature and applies L7 networkPolicy. If user wants to visualize the L7 flows, user may enable `L7Visibility` and does not have to annotate the pods included in the L7 networkPolicy applied. 
if user does not enable `L7NetworkPolicy` but `L7Visibility`, only pods with L7 visibility annotation will be in the flow table with Layer 7 flow information.

- Single Mechanism: 
Regardless of `L7NetworkPolicy` feature gate enabled, we will only update the tables if the Pods or Namespace has L7 visibility annotation and `L7Visibility` feature gate enabled. The only redundancy I see here is the flow which is already flowing through Suricata (in case of L7 NP) is mirrored again to the Suricata interface using TC mark flows.

Reply:
I have made the change in the exporter.go file for only accepting L7 flows if L7Visibility is enabled

Reply:
> The only redundancy I see here is the flow which is already flowing through Suricata (in case of L7 NP) is mirrored again to the Suricata interface using TC mark flows.

I think I may be missing something. Why can't we send the flow to Suricata _exactly once_, as long as at least one of the following conditions is met:
- an L7NP is applied to the Pod (and the L7NP FeatureGate is enabled)
- the Pod or its Namespace is annotated with the visibility information (and the L7Visibility FeatureGate is enabled)

Reply:
Current implementation is the same with just one change
 
> an L7NP is applied to the Pod (and the L7NP FeatureGate is enabled) 
and the flows appear in the flow tables if L7Visibility FeatureGate is enabled

Reply:
I think we should clearly differentiate between what the implementation does, and what the user-facing behavior is.

From a user perspective, L7NetworkPolicy and L7Visibility should be independent features. There should be no dependency on L7NetworkPolicy for L7Visibility, and the L7Visibility feature should behave the same whether L7NetworkPolicy is enabled or not.

From an implementation perspective, we should do whatever makes the most sense (performance and simplicity-wise) to achieve the desired user-facing behavior.

Reply:
The two are independent features. I think the confusion was due to the enabling of L7listener if L7NetworkPolicy is enabled, which is now changed to L7Visibility in exporter.go.

Reply:
After discussion with @antoninbas, new patch has been added which only exports http flows to the Flow Aggregator iff there is a pod/namespace annotation is present.
Following scenarios are possible:
- Only L7NetworkPolicy is enabled: No flow containing http fields will be received by the exporter.
- L7NetworkPolicy is enabled and L7Visibility is also enabled, however, Pod/Namespace is not annotated:
Flows will be received by the exporter but will be discarded as the corresponding Pod is not annotated.
- L7Visibility enabled and Pods/Namespace annotated:
Flows will be received and on query will be send to Flow Aggregator.
This behavior will be followed irrespective of L7NetworkPolicy enable/disable status.

Suricata will utilize only one instance shared by both reconciler and L7VisibilityController, and will start only once.